create_exception: the trait bound `MyError: pyo3::PyNativeType` is not satisfied

### Bug Description

while migrating to pyo3 0.22.3 the declared exceptions were no longer compiling due to a missing trait bound

### Steps to Reproduce

1. define a `create_exception!(my_module, MyError, PyException, "Some description.");`
2. it doen't compile

### Backtrace

```shell
error[E0277]: the trait bound `MyError: pyo3::PyNativeType` is not satisfied
  --> \bindings\python\src\exceptions.rs:4:1
   |
4  | create_exception!(my_module, MyError, PyException, "Some description.");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `pyo3::PyNativeType` is not implemented for `MyError`
   |
   = help: the following other types implement trait `pyo3::PyNativeType`:
             PyArray<T, D>
             PyArrayDescr
             PyUntypedArray
             PyCell<T>
             PyAny
             PyBool
             PyByteArray
             PyBytes
           and 115 others
   = note: required for `MyError` to implement `HasPyGilRef`
note: required by a bound in `PyTypeInfo`
  --> .cargo\registry\src\index.crates.io-6f17d22bba15001f\pyo3-0.22.3\src\type_object.rs:63:38
   |
63 | pub unsafe trait PyTypeInfo: Sized + HasPyGilRef {
   |                                      ^^^^^^^^^^^ required by this bound in `PyTypeInfo`
   = note: this error originates in the macro `$crate::create_exception_type_object` which comes from the expansion of the macro `create_exception` (in Nightly builds, run with -Z macro-backtrace for more info)
```


### Your operating system and version

Windows 10

### Your Python version (`python --version`)

3.12

### Your Rust version (`rustc --version`)

1.75

### Your PyO3 version

0.22

### How did you install python? Did you use a virtualenv?

python is not required to reproduce the issue

### Additional Info

_No response_

--- a/src/types/mod.rs
+++ b/src/types/mod.rs
@@ -121,14 +121,24 @@ pub trait DerefToPyAny {
 // Implementations core to all native types
 #[doc(hidden)]
 #[macro_export]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+#[cfg(not(feature = "gil-refs"))]
+macro_rules! pyobject_native_type_base(
+    // empty implementation on non-gil-refs
+    ($name:ty $(;$generics:ident)* ) => {};
+);
+
+// Implementations core to all native types
+#[doc(hidden)]
+#[macro_export]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+#[cfg(feature = "gil-refs")]
 macro_rules! pyobject_native_type_base(
     ($name:ty $(;$generics:ident)* ) => {
-        #[cfg(feature = "gil-refs")]
         unsafe impl<$($generics,)*> $crate::PyNativeType for $name {
             type AsRefSource = Self;
         }
 
-        #[cfg(feature = "gil-refs")]
         impl<$($generics,)*> ::std::fmt::Debug for $name {
             fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>)
                    -> ::std::result::Result<(), ::std::fmt::Error>
@@ -139,7 +149,6 @@ macro_rules! pyobject_native_type_base(
             }
         }
 
-        #[cfg(feature = "gil-refs")]
         impl<$($generics,)*> ::std::fmt::Display for $name {
             fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>)
                    -> ::std::result::Result<(), ::std::fmt::Error>
@@ -157,7 +166,6 @@ macro_rules! pyobject_native_type_base(
             }
         }
 
-        #[cfg(feature = "gil-refs")]
         impl<$($generics,)*> $crate::ToPyObject for $name
         {
             #[inline]
@@ -172,6 +180,43 @@ macro_rules! pyobject_native_type_base(
 // make sense on PyAny / have different implementations).
 #[doc(hidden)]
 #[macro_export]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+#[cfg(not(feature = "gil-refs"))]
+macro_rules! pyobject_native_type_named (
+    ($name:ty $(;$generics:ident)*) => {
+
+        impl<$($generics,)*> ::std::convert::AsRef<$crate::PyAny> for $name {
+            #[inline]
+            fn as_ref(&self) -> &$crate::PyAny {
+                &self.0
+            }
+        }
+
+        impl<$($generics,)*> ::std::ops::Deref for $name {
+            type Target = $crate::PyAny;
+
+            #[inline]
+            fn deref(&self) -> &$crate::PyAny {
+                &self.0
+            }
+        }
+
+        unsafe impl<$($generics,)*> $crate::AsPyPointer for $name {
+            /// Gets the underlying FFI pointer, returns a borrowed pointer.
+            #[inline]
+            fn as_ptr(&self) -> *mut $crate::ffi::PyObject {
+                self.0.as_ptr()
+            }
+        }
+
+        impl $crate::types::DerefToPyAny for $name {}
+    };
+);
+
+#[doc(hidden)]
+#[macro_export]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+#[cfg(feature = "gil-refs")]
 macro_rules! pyobject_native_type_named (
     ($name:ty $(;$generics:ident)*) => {
         $crate::pyobject_native_type_base!($name $(;$generics)*);
@@ -202,7 +247,6 @@ macro_rules! pyobject_native_type_named (
 
         // FIXME https://github.com/PyO3/pyo3/issues/3903
         #[allow(unknown_lints, non_local_definitions)]
-        #[cfg(feature = "gil-refs")]
         impl<$($generics,)*> $crate::IntoPy<$crate::Py<$name>> for &'_ $name {
             #[inline]
             fn into_py(self, py: $crate::Python<'_>) -> $crate::Py<$name> {
@@ -212,7 +256,6 @@ macro_rules! pyobject_native_type_named (
 
         // FIXME https://github.com/PyO3/pyo3/issues/3903
         #[allow(unknown_lints, non_local_definitions)]
-        #[cfg(feature = "gil-refs")]
         impl<$($generics,)*> ::std::convert::From<&'_ $name> for $crate::Py<$name> {
             #[inline]
             fn from(other: &$name) -> Self {
@@ -223,7 +266,6 @@ macro_rules! pyobject_native_type_named (
 
         // FIXME https://github.com/PyO3/pyo3/issues/3903
         #[allow(unknown_lints, non_local_definitions)]
-        #[cfg(feature = "gil-refs")]
         impl<'a, $($generics,)*> ::std::convert::From<&'a $name> for &'a $crate::PyAny {
             fn from(ob: &'a $name) -> Self {
                 unsafe{&*(ob as *const $name as *const $crate::PyAny)}
@@ -279,11 +321,23 @@ macro_rules! pyobject_native_type_info(
 // because rust-numpy has a special implementation.
 #[doc(hidden)]
 #[macro_export]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+#[cfg(not(feature = "gil-refs"))]
+macro_rules! pyobject_native_type_extract {
+    // no body for non-gil-refs
+    ($name:ty $(;$generics:ident)*) => {};
+}
+
+// NOTE: This macro is not included in pyobject_native_type_base!
+// because rust-numpy has a special implementation.
+#[doc(hidden)]
+#[macro_export]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+#[cfg(feature = "gil-refs")]
 macro_rules! pyobject_native_type_extract {
     ($name:ty $(;$generics:ident)*) => {
         // FIXME https://github.com/PyO3/pyo3/issues/3903
         #[allow(unknown_lints, non_local_definitions)]
-        #[cfg(feature = "gil-refs")]
         impl<'py, $($generics,)*> $crate::FromPyObject<'py> for &'py $name {
             #[inline]
             fn extract_bound(obj: &$crate::Bound<'py, $crate::PyAny>) -> $crate::PyResult<Self> {

--- a/src/types/mod.rs
+++ b/src/types/mod.rs
@@ -201,14 +201,6 @@ macro_rules! pyobject_native_type_named (
             }
         }
 
-        unsafe impl<$($generics,)*> $crate::AsPyPointer for $name {
-            /// Gets the underlying FFI pointer, returns a borrowed pointer.
-            #[inline]
-            fn as_ptr(&self) -> *mut $crate::ffi::PyObject {
-                self.0.as_ptr()
-            }
-        }
-
         impl $crate::types::DerefToPyAny for $name {}
     };
 );

